octobot_commons/databases/cache_manager.py

```diff
@@ -0,0 +1,205 @@
+        cache_full_path = self.get_cache_path(tentacle, exchange, symbol, time_frame,
+                                              tentacle_name, tentacles_setup_config)
+        cache_dir = os.path.split(cache_full_path)[0]
+        if not os.path.exists(cache_dir):
+            os.makedirs(cache_dir)
+        return self._open_cache_database(cache_full_path, cache_type)
+
+    def _open_cache_database(self, file_path, cache_type):
+        """
+        Override to use another cache database or adaptor
+        :return: the cache database class
+        """
+        return _CacheWrapper(file_path, cache_type, self.database_adaptor)
+
+    def get_cache_path(self, tentacle, exchange, symbol, time_frame, tentacle_name, tentacles_setup_config):
+        try:
+            return self.__class__.CACHES[tentacle_name][exchange][symbol][time_frame].get_path()
+        except KeyError:
+            sanitized_pair = symbol_util.merge_symbol(symbol) if symbol else symbol
+            # warning: very slow, should be called as rarely as possible
+            required_tentacles = self._get_linked_tentacles(tentacle, tentacles_setup_config)
+            identifying_tentacles = [tentacle] + required_tentacles
+            return os.path.join(common_constants.USER_FOLDER, common_constants.CACHE_FOLDER, tentacle_name,
+                                exchange, sanitized_pair, time_frame,
+                                self._code_hash(identifying_tentacles),
+                                self._config_hash(identifying_tentacles, tentacles_setup_config),
+                                common_constants.CACHE_FILE)
+
+    def _get_linked_tentacles(self, tentacle, tentacles_setup_config):
+        try:
+            import octobot_tentacles_manager.api as tentacles_manager_api
+            linked_tentacles_classes = tentacles_manager_api.get_tentacle_classes_requirements(tentacle)
+            tentacles = []
+            for tentacle_class in linked_tentacles_classes:
+                # try instantiating tentacles, if not possible, then ignore it. Works with evaluator tentacles.
+                try:
+                    tentacles.append(tentacle_class(tentacles_setup_config))
+                except TypeError:
+                    pass
+            return tentacles
+        except ImportError as e:
+            raise ImportError("octobot_tentacles_manager is required to use cache") from e
+
+    def _code_hash(self, identifying_tentacles) -> str:
+        full_code = ""
+        for linked_tentacle in identifying_tentacles:
+            code_location = linked_tentacle.get_script() if hasattr(linked_tentacle, "get_script") \
+                else linked_tentacle.__class__
+            tentacle_code = inspect.getsource(code_location).replace(" ", "").replace("\n", "")
+            full_code = f"{full_code}{tentacle_code}"
+        return hashlib.sha256(full_code.encode()).hexdigest()[:common_constants.CACHE_HASH_SIZE]
```

Comment:
Why trucate the sha265 result? For size reasons?

Reply:
Yes, to avoid 256-char folder names, if you have a better idea, we change this :)